sdk/iot/pnp/inc/az_iot_pnp_client.h

```diff
@@ -0,0 +1,332 @@
+    az_iot_pnp_client const* client,
+    uint32_t token_expiration_epoch_time,
+    az_span signature,
+    az_span* out_signature);
+
+/**
+ * @brief Gets the MQTT password.
+ * @note The MQTT password must be an empty string if X509 Client certificates are used. Use this
+ *       API only when authenticating with SAS tokens.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] base64_hmac_sha256_signature The Base64 encoded value of the HMAC-SHA256(signature,
+ *                                         SharedAccessKey). The signature is obtained by using
+ *                                         #az_iot_pnp_client_sas_signature_get.
+ * @param[in] key_name The Shared Access Key Name (Policy Name). This is optional. For security
+ *                     reasons we recommend using one key per device instead of using a global
+ *                     policy key.
+ * @param[in] mqtt_password An empty #az_span with sufficient capacity to hold the MQTT password.
+ * @param[out] out_mqtt_password The output #az_span containing the MQTT password.
+ * @return #az_result.
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_sas_password_get(
+    az_iot_pnp_client const* client,
+    az_span base64_hmac_sha256_signature,
+    az_span key_name,
+    az_span mqtt_password,
+    az_span* out_mqtt_password);
+
+/*
+ *
+ * PnP Telemetry APIs
+ *
+ */
+
+/**
+ * @brief Azure IoT PnP telemetry options.
+ *
+ */
+typedef struct az_iot_pnp_client_telemetry_options
+{
+  struct
+  {
+    // OPEN: I can't find samples in Azure SDK modeling this.  Closest is
+    // sdk\storage\blobs\inc\az_storage_blobs.h where it has 'az_span option" in
+    // az_storage_blobs_blob_download_options that's ignored in practice.
+    int _not_used;
+  } _internal;
+} az_iot_pnp_client_telemetry_options;
+
+/**
+ * @brief Gets the default Azure IoT PnP telemetry options.
+ * @details Call this to obtain an initialized #az_iot_pnp_client_telemetry_options structure that
+ * can be afterwards modified and passed to #az_iot_pnp_client_telemetry_publish_topic_get.
+ *
+ * @return #az_iot_pnp_client_telemetry_options.
+ */
+AZ_NODISCARD az_iot_pnp_client_telemetry_options az_iot_pnp_client_telemetry_options_default();
+
+/**
+ * @brief Gets the MQTT topic that must be used for device to cloud telemetry messages.
+ * @note Telemetry MQTT Publish messages must have QoS At Least Once (1).
+ * @note This topic can also be used to set the MQTT Will message in the Connect message.
+ *
+ * Should the user want a null terminated topic string, they may allocate a buffer large enough
+ * to fit the topic plus a null terminator. They must set the last byte themselves or zero
+ * initialize the buffer.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] component_name An #az_span specifying the component name to publish telemetry on.
+ * @param[in] content_type An #az_span specifying the content type of the message payload.
```

Comment:
will the available options for this be documented somewhere or is it "anything goes"

Reply:
There's rules in practice about what various backends can be expected to support (basically utf-8 and application/json).  I'm putting notes of what's currently supported inline just to steer folks correctly.